### PR TITLE
Eren talon pid

### PR DESCRIPTION
**Pull request by Ergoold**

### CODE REVIEW

**Comment by Ergoold** [comment]
[Add A PIDLoop Interface](https://trello.com/c/JU0zMz21/11-add-a-pidloop-interface)

**Comment by Ergoold** [comment]
[Implement TalonSRX PID Loops Using PIDLoop](https://trello.com/c/k7tBqDVM/12-implement-talonsrx-pid-loops-using-pidloop)

File: src/main/java/com/spikes2212/control/TalonPIDLoop.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+     * 0 - primary loop
+     * 1 - secondary loop
+     */
+    private int loop;
+
+    private int timeout;
+
+    private double lastTimeNotOnTarget;
+
+    public TalonPIDLoop(BaseMotorController motor, Supplier<Double> kp, Supplier<Double> ki, Supplier<Double> kd,
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/control/TalonPIDLoop.java]: rename parameters to `kP`

File: src/main/java/com/spikes2212/control/TalonPIDLoop.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        motor.set(ControlMode.PercentOutput, 0);
+    }
+
+    @Override
+    public void update() {
+        motor.config_kP(loop, kp.get(), timeout);
+        motor.config_kI(loop, ki.get(), timeout);
+        motor.config_kD(loop, kd.get(), timeout);
+
+        motor.set(ControlMode.Position, setpoint.get());
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/control/TalonPIDLoop.java]: receive control mode as a parameter to the Loop

File: src/main/java/com/spikes2212/control/TalonPIDLoop.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        this.lastTimeNotOnTarget = Timer.getFPGATimestamp();
+    }
+
+    private void initialize() {
+        motor.configFactoryDefault();
+        motor.configSelectedFeedbackSensor(FeedbackDevice.CTRE_MagEncoder_Relative, loop, timeout);
+
+        motor.configNominalOutputForward(0, timeout);
+        motor.configNominalOutputReverse(0, timeout);
+        motor.configPeakOutputForward(1.0, timeout);
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/control/TalonPIDLoop.java]: receive absolute `PeakOutput` as a property in the constructor

File: src/main/java/com/spikes2212/control/TalonPIDLoop.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        motor.configPeakOutputReverse(-1.0, timeout);
+
+        motor.configAllowableClosedloopError(loop, 0, timeout);
+
+        motor.config_kP(loop, kP.get(), timeout);
+        motor.config_kI(loop, kI.get(), timeout);
+        motor.config_kD(loop, kD.get(), timeout);
+
+        motor.setSelectedSensorPosition(0, loop, timeout);
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/control/TalonPIDLoop.java]: remove this line, the loop wouldn't start at 0 necessarily

File: src/main/java/com/spikes2212/control/TalonPIDLoop.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        motor.config_kP(loop, kP.get(), timeout);
+        motor.config_kI(loop, kI.get(), timeout);
+        motor.config_kD(loop, kD.get(), timeout);
+
+        motor.set(controlMode, setpoint.get());
+    }
+
+    @Override
+    public boolean onTarget() {
+        boolean currentlyOnTarget = Math.abs(setpoint.get() - motor.getSelectedSensorPosition(loop)) < tolerance.get();
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/control/TalonPIDLoop.java]: this wouldn't neccesserally use `getSelectedSensorPosition` please find a way to make this more generic

File: src/main/java/com/spikes2212/control/TalonPIDLoop.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+    /**
+     * The acceptable distance from the target.
+     */
+    private Supplier<Double> tolerance;
+
+    /**
+     * The `canMove` method of the Subsystem this PIDLoop belongs to.
+     */
+    private Function<Double, Boolean> canMove;
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/control/TalonPIDLoop.java]: replace `Function<Double,Boolean>` with `Predicate<Double>`.

**Comment by Ergoold** [comment]
We are waiting until #9 is merged into this before merging to `dev`, right?

**Pull request merged**